articles/remapping.md

```diff
@@ -0,0 +1,119 @@
+### Remapping Names
+#### ROS 1
+In ROS 1 [Names](http://wiki.ros.org/Names) can be remapped to other names by passing [remapping arguments](http://wiki.ros.org/Remapping%20Arguments) via the command line.
+`roscpp` additionally allows remaps to be passed via `init()` as strings in STL containers.
+A ROS1 remap consists of two names: one that should be replaced with another.
+
+ROS 1 remapping works on **Fully Qualified Names** (FQN).
+`roscpp` [names::init()](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#a377ff8fede7b95398fd5d1c5cd49246b) expands both sides of a remap to FQNs before storing it.
+[names::remap()](http://docs.ros.org/api/roscpp/html/namespaceros_1_1names.html#ab2eebaf734abfbdccb4122f8e24f547f) expands a name to a FQN before checking for applicable remapping rules.
+The name is remapped to the right side only if it exactly matches the left side of a rule.
+
+#### ROS 2
+In ROS2 remapping will be more powerful.
+See the requirements section below for examples.
+There are more types of remappings that will be supported:
+
+- Namespace replacement
+
+ - Replace all or part of a namespace with another namespace
+ 
+- Basename replacement
+
+ - Replace all of the basename with another basename
+ 
+- Arbitrary substring replacement ***(maybe)***
+
+ - Replace any substring with another string
+ 
+- Exact Name replacement
+
+ - Replace an entire name with another name
+ 
+ - *ROS 1 remapping is this case applied post-FQN expansion*
+
+ROS 2 may support remapping rules being applied either before or after the names are expanded to FQN.
+ROS 1 only supports the latter.
+
+#### Pre-FQN expansion remapping
+Say the source code for a node use two names `cat` and `/ns/cat`.
+If the node runs in namespace `ns`, the FQN of both topics will be identical: `/ns/cat`.
+No rule could remap these names individually once they have been expanded.
+A rule applied Pre-FQN expansion could remap just one of them.
+
+#### Requirements for remapping capabilities
+##### 1. Namespace Replacement
+These are rules that only apply to the namespace part of the name.
```

Comment:
Is it really different that we have different remapping types? Since this is a list of requirements it shouldn't presume how to implement this.

The namespace/basename distinction seems to collapse if we support filesystem like regexes. But the requirement to support both use cases is still valid.